### ISSUE TITLE
Min/Max per block. Querying the min() and max() of the primary key.

### ISSUE THREAD

**Issue by fjolliton** [yandex/ClickHouse#329]
We are doing some tests on ClickHouse on a small database (~150 GiB, 10 tables, less than 1 billion rows in each table, single node, `Buffer` + `MergeTree`, 1.1.54127).

Our primary key is a single `UInt64` column (a timestamp actually).

When querying the `min()` and `max()` of this column, it seems that ClickHouse scan it entirely. Isn't that a good case for using the index to get an O(1) response?

Also, I was wondering if this is due to the implementation and I was thinking about something that could make the query faster for columns that doesn't contain too much "randomness": storing the minimum and maximum value in the header of each blocks would allow to quickly skip them in some circonstances. What do you think about that?

**Comment by ludv1x** [comment]
Yes, ClickHouse makes full scan of all MergeTree's parts and applies aggregate function for the column.

This optimization (if column is PK) is possible now even without need to add metainformation for blocks, since MergeTree stores values in sorted order. So it is enough to read first and last row of the column (or primary key index).

Your proposed solution assumes storing as many as necessary metadata (min, max, count, avg. size) for columns in special files inside MergeTree's parts (i.e. blocks). It is good idea. We discussed such approach, it is really more common and flexible and allows to do many other sophisticated optimizations. But we don't plan to implement it in near future.

Currently we are modifying ExpressionAnalyzer which is responsible to construct query graph form query AST.
This modification will allow to add such "particular case" optimizations more easily.
After this work will be finished, we plan to add certain "particular" optimizations and we will consider your optimization too.

@alexey-milovidov

**Comment by innerr** [comment]
Is this feature deliver yet?

A further question: any plan of storing min/max values for each 8k-rows of a column?

**Comment by alexey-milovidov** [comment]
This is still in our longlist.

**Comment by orloffv** [comment]
Any news?

**Comment by alexey-milovidov** [comment]
https://github.com/yandex/ClickHouse/pull/3973

**Comment by alexey-milovidov** [comment]
This should be similar to the "trivial count optimization" but for min/max.
We can extend "trivial count optimization" for queries without GROUP BY and without filter (or with filter by partition) that contain only count, min and max aggregate functions.
